aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -478,5 +490,182 @@ std::tuple<Tensor, Tensor> native_multi_head_attention(
+  TORCH_CHECK(
+      D == qkv_weight.sizes()[1],
+      "expected `qkv_weight` second dim to be embed_Dim");
+  TORCH_CHECK(
+      qkv_bias.dim() == 1,
+      "expected 2-D `qkv_bias`, got ",
+      qkv_bias.dim(),
+      "-D tensor");
+  TORCH_CHECK(
+      qkv_bias.sizes()[0] == 3 * D,
+      "expected `qkv_bias` first dim and first dim of query to be equal");
+  TORCH_CHECK(D % num_head == 0, "`embed_dim` must divide evenly by `num_heads`");
+
+#ifndef NDEBUG
+  const auto B = query.is_nested()
+      ? get_nested_tensor_impl(query)->get_nested_size_tensor().size(0)
+      : query.sizes()[0];
+  auto T = query.is_nested() ? 0 : query.sizes()[1];
+  const auto dim_per_head = D / num_head;
+#endif
+
+  // shape: [B, T, 3 x D]
+  const bool is_incremental_decoding = incr_key.has_value() && incr_value.has_value() && incr_key->sizes() == incr_value->sizes();
+  auto qkv = qkv_matmul(query, query, query, embed_dim, qkv_weight);
+
+  if (!qkv.is_nested() && qkv.numel() == 0) {
+    if (query.is_nested()) {
+      return std::make_tuple(Tensor(), Tensor(), Tensor(), Tensor());
+    }
+    return std::make_tuple(at::empty_like(query), Tensor(), Tensor(), Tensor());
+  }
+
+#ifndef NDEBUG
+  if (!query.is_nested() || !qkv.is_nested()) {
+    if (query.is_nested()) {
+      T = qkv.size(1);
+    }
+    debug_assert_shape(__LINE__, qkv, {B, T, 3 * D});
+  }
+#endif
+
+#ifdef DEBUG_PRINT_EACH_STEP
+  if (!qkv.is_nested()) {
+    std::cerr << "qkv: " << qkv << std::endl;
+  }
+#endif
+  // shape: 3 x [B, num_head, T, dim_per_head]
+  auto q_k_v = _transform_bias_rescale_qkv(qkv, qkv_bias, num_head);
+  qkv = Tensor(); // Not used any more, allow free
+  auto& q = std::get<0>(q_k_v);
+  const auto& _k = std::get<1>(q_k_v);
+  const auto& _v = std::get<2>(q_k_v);
+#ifndef NDEBUG
+  debug_assert_shape(__LINE__, q, {B, num_head, T, dim_per_head});
+  debug_assert_shape(__LINE__, _k, {B, num_head, T, dim_per_head});
+  debug_assert_shape(__LINE__, _v, {B, num_head, T, dim_per_head});
+#endif
+#ifdef DEBUG_PRINT_EACH_STEP
+  std::cerr << "q: " << q << std::endl;
+  std::cerr << "k: " << _k << std::endl;
+  std::cerr << "v: " << _v << std::endl;
+#endif
+  Tensor k;
+  Tensor v;
+  if (is_incremental_decoding) {
+    // TODO: is this actually setting incr_key and incr_value in place
+    // what if k and incr k are nestedtensors? same for v case.
+    k = at::cat({incr_key.value(), _k}, 2);
+    v = at::cat({incr_value.value(), _v}, 2);
+  } else {
+    k = _k;
+    v = _v;
+  }
+  // shape: [B, num_head, T, T]
+  auto qkt = bmm_nt(q, k);
+  // q & k are dead but cannot be freed because they were packed with v
+#ifndef NDEBUG
+  debug_assert_shape(__LINE__, qkt, {B, num_head, T, T});
+#endif
+#ifdef DEBUG_PRINT_EACH_STEP
+  std::cerr << "qkt: " << qkt << std::endl;
+#endif
+
+  // shape: [B, num_head, T, T]
+  // TODO: long-term, have a kernel that works with
+  // NestedTensor directly if there is no mask passed
+  qkt = masked_softmax(qkt, mask, query);
```

Comment:
Do we even have nested masedk softmax already?

Reply:
I copied this from transformer_encoder_layer_fwd, so I don't believe so. On testing, would you recommend unit tests that just check for invalid shapes like so https://github.com/pytorch/pytorch/blob/543919cfc8d7014dfa365b55c459fa2b3f746891/test/test_nn.py#L5878? I don't see any numerical testing for transformer_encoder_layer_fwd. 

I have numerical testing in an internal benchmarking script where I compare against fairseq, but putting that test in torch.testing would mean I need to import fairseq in the test, which isn't a reliable dependency. I can't compare against nn.TransformerDecoder because we don't have an existing decoder-only layer, only encoder-decoder.

Reply:
Tests are in test_transformers.py here https://github.com/pytorch/pytorch/pull/79576.